### PR TITLE
Allow proper mapping of tenant hostnames to URLs

### PR DESCRIPTION
**Pull request by mtneug**
Opencast has some assumptions about how hosts are accessed in different
organizations. Especially the schema and port number are assumed to be
equal or in certain cases just the default protocol port number.

This changes the host mapping configuration to explicitly specify a URL.

### Your pull request should…

* [x] have a concise title
* [x] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [x] be against the correct branch (features can only go into develop)
* [x] include migration scripts and documentation, if appropriate
* [x] pass automated tests
* [x] have a clean commit history
* [x] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

### CODE REVIEW

File: etc/org.opencastproject.organization-mh_default_org.cfg [hunk 1]
```diff
@@ -52,26 +52,17 @@ name=Opencast Project
 #
 # Example Configuration:
 #
-# prop.org.opencastproject.host.admin-presentation.opencast.com=tenant1.admin-presentation.opencast.com
-# prop.org.opencastproject.host.ingest.opencast.com=tenant1.ingest.opencast.com
-# prop.org.opencastproject.host.worker.opencast.com=tenant1.worker.opencast.com
+# prop.org.opencastproject.host.admin-presentation.opencast.com=http://tenant1.admin-presentation.opencast.com
+# prop.org.opencastproject.host.ingest.opencast.com=https://tenant1.ingest.opencast.com
```
Comment (lkiesow) [inline etc/org.opencastproject.organization-mh_default_org.cfg:56]: Does this mean that you cannot have a tenant that reacts to both http and https any longer?

File: modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java [hunk 2]
```diff
@@ -164,7 +164,11 @@ public static String getSystemUserName(ComponentContext cc) {
 
   /** Extract hostname and port number from a URL. */
   public static Tuple<String, Integer> hostAndPort(URL url) {
-    return tuple(StringUtils.strip(url.getHost(), "/"), url.getPort());
+    int port = url.getPort();
+    if (port < 0) {
```
Comment (lkiesow) [inline modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java:168]: Previously, `port=-1` meant any port. Does this now mean something else?

**Review by lkiesow — COMMENTED** [review]
Haven't really looked at this closely. Just some questions…

File: modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java (hunk 2 above)
Comment (mtneug) [reply modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java:168]: When a URL is parsed, `-1` means that no port is explicitly specified in the URL. We can also check for `-1`, but every port number below 0 is invalid.

File: etc/org.opencastproject.organization-mh_default_org.cfg (hunk 1 above)
Comment (mtneug) [reply etc/org.opencastproject.organization-mh_default_org.cfg:56]: No, if I understand the code correctly.

The schema only matters for the `ServiceRegistryEndpoint` `/available.{xml|json}`. pyCA, for instance, should then only use the schema and port specified here. Note that the other service registry endpoints will still report the URL of the service registration (i.e. `org.opencastproject.server.url`). I would expect them to be tenant-specific as well, but I don't know if this is intentionally limited to this one endpoint.

Opencast currently determines the tenant using only the hostname. You should be able to route traffic through different ports or a different protocol schema. As mentioned in the updated docs, this also means that it is currently not possible to have `tennen1` use `http://localhost:8080` and `tennen2` `http://localhost:8081`.

So including a schema here is necessary for the `ServiceRegistryEndpoint` and to not explicitly specifying a port number.

Comment (mtneug) [reply etc/org.opencastproject.organization-mh_default_org.cfg:56]: mmh I was wrong. The port is used to differentiate between tenants. Will update the docs.

File: modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java (hunk 2 above)
Comment (mtneug) [reply modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java:168]: Can this be resolved?

Comment (mliradelc) [reply modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java:168]: @lkiesow is this solved for you? 🙂

Comment (mtneug) [reply modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java:168]: ?

**Review by gregorydlogan — APPROVED** [review]
LGTM, but I also don't have a multitenant system to test with...

File: modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java (hunk 2 above)
Comment (lkiesow) [reply modules/common/src/main/java/org/opencastproject/security/util/SecurityUtil.java:168]: Don't wait for me. If this is fine, it's fine. It was just something I noticed. I didn't do a proper review of this.

**Comment by lkiesow** [comment]
@mtneug, can you add a release notes snippet for this?
See https://github.com/opencast/opencast/tree/develop/docs/guides/admin/docs/releasenotes

**Pull request merged**